### PR TITLE
Api challenge

### PR DESCRIPTION
**Pull request by nhrin**

### CODE REVIEW

File: api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import java.time.Period;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Service
+public class DeIdentificationPatientRecordService {
+
+    private final String EMAIL_REGEX = "[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\\.[a-zA-Z0-9-.]+";
```
Comment (FedirKhodko) [inline api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java]: static for all consts

File: api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    private final String US_SOCIAL_SECURITY_NUMBER_REGEX = "\\d{3}-?\\d{2}-?\\d{4}";
+    private final String US_TELEPHONE_NUMBER_REGEX = "(\\+\\d{1,3}( )?)?((\\(\\d{3}\\))|\\d{3})[- .]?\\d{3}[- .]?\\d{4}";
+    private final Map<String, Integer> DATE_REGEX = new HashMap<>() {{
+        put("\\s+(?:19\\d{2}|20\\d{2})[-/.](?:0[1-9]|1[012])[-/.](?:0[1-9]|[12][0-9]|3[01])\\b", 1); //YYYY/MM/DD
+        put("\\s+(?:0[1-9]|[12][0-9]|3[01])[-/.](?:0[1-9]|1[012])[-/.](?:19\\d{2}|20\\d{2})\\b", 7); //DD/MM/YYYY
+        put("\\s+(?:0[1-9]|1[012])[-/.](?:0[1-9]|[12][0-9]|3[01])[-/.](?:19\\d{2}|20\\d{2})\\b", 7); //MM/DD/YYYY
+    }};
+
+
+    public JSONObject deIdentificationAllInfo(Patient patient) {
```
Comment (FedirKhodko) [inline api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java]: It's better to start method names with a verb. Like getDeiden...,  prepareDeiden..

**Comment by FedirKhodko** [comment]
Why lambda related files are also here?

**Comment by FedirKhodko** [comment]
Lets add some unit tests here.

**Comment by nhrin** [comment]
> Why lambda related files are also here?

There are not lambda related files. I just moved them into another directory.

File: api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java (hunk 1 above)
Comment (nhrin) [reply api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java]: fixed

File: api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java (hunk 2 above)
Comment (nhrin) [reply api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java]: fixed

**Comment by nhrin** [comment]
> Lets add some unit tests here.

added

File: api-challenge/src/test/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringBootTest
+public class DeIdentificationPatientRecordServiceTest {
+
+    @Autowired
+    private DeIdentificationPatientRecordService service;
+
+    @Test
+    public void testCalculateAge() {
```
Comment (FedirKhodko) [inline api-challenge/src/test/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordServiceTest.java]: Usually we call test methods using BDD approach: shouldReturn...., shouldReplace..., shouldThrow...,

File: api-challenge/src/test/java/nix/nhrin/apichallenge/ApiChallengeApplicationTests.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package nix.nhrin.apichallenge;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class ApiChallengeApplicationTests {
```
Comment (FedirKhodko) [inline api-challenge/src/test/java/nix/nhrin/apichallenge/ApiChallengeApplicationTests.java]: Do we need this file?

File: api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                String year = forReplace.substring(entry.getValue(), entry.getValue() + 4);
+                result = result.replaceAll(forReplace, " " + year + " year");
+            }
+        }
+        return result;
+    }
+
+    @SneakyThrows
+    public String getDeIdentificationZipCode(String zipCode) {
+        try (CSVReader csvReader = new CSVReader(new FileReader("src/main/resources/population_by_zcta_2010.csv"))) {
```
Comment (FedirKhodko) [inline api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java]: I'd prefer to load this table to memory during object creation. Can place that data to map data structure. Thus we will have O(1) complexity of retrieving zip code related data.

File: api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.web.bind.annotation.ResponseBody;
+
+@Controller
+public class PatientController {
+
+    @Autowired
+    private DeIdentificationPatientRecordService deIdentificationPatientRecordService;
+
+    @ResponseBody
+    @PostMapping("/deIdentification")
```
Comment (FedirKhodko) [inline api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java]: Let call it deidentificate

File: api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    public JSONObject deIdentificationPatientRecord (@RequestBody Patient patient) {
```
Comment (FedirKhodko) [inline api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java]: deidentificatePatientRecord

File: api-challenge/src/test/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordServiceTest.java (hunk 3 above)
Comment (nhrin) [reply api-challenge/src/test/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordServiceTest.java]: renamed

File: api-challenge/src/test/java/nix/nhrin/apichallenge/ApiChallengeApplicationTests.java (hunk 4 above)
Comment (nhrin) [reply api-challenge/src/test/java/nix/nhrin/apichallenge/ApiChallengeApplicationTests.java]: removed

File: api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java (hunk 5 above)
Comment (nhrin) [reply api-challenge/src/main/java/nix/nhrin/apichallenge/service/DeIdentificationPatientRecordService.java]: fixed

File: api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java (hunk 6 above)
Comment (nhrin) [reply api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java]: fixed

File: api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java (hunk 7 above)
Comment (nhrin) [reply api-challenge/src/main/java/nix/nhrin/apichallenge/controller/PatientController.java]: fixed

**Pull request merged**